param/__init__.py

```diff
@@ -1287,6 +1287,112 @@ def __init__(self, **params):
         super(Array,self).__init__(ndarray, allow_None=True, **params)
 
 
+class DataFrame(ClassSelector):
+    """
+    Parameter whose value is a pandas DataFrame.
+    """
+    __slots__ = ['rows','columns', 'ordered']
+
+    def __init__(self, rows=None, columns=None, ordered=None, **params):
+        from pandas import DataFrame as pdDFrame
+        self.rows = rows
+        self.columns = columns
+        self.ordered = ordered
+        super(DataFrame,self).__init__(pdDFrame, allow_None=True, **params)
+        self._check_value(self.default)
+
+
+    def _length_bounds_check(self, bounds, length, name):
+        message = '{name} length {length} does not match declared bounds of {bounds}'
+        if not isinstance(bounds, tuple):
+            if (bounds != length):
+                raise Exception(message.format(name=name, length=length, bounds=bounds))
+            else:
+                return
+        (lower, upper) = bounds
+        if lower is not None:
+            if length < lower:
+                raise Exception(message.format(name=name, length=length, bounds=bounds))
+        if upper is not None:
+            if length > upper:
+                raise Exception(message.format(name=name,length=length, bounds=bounds))
```

Comment:
Probably clearer if expressed a single if statement, with one test expression?

Reply:
Addressed in 70642bc